python/ray/rllib/optimizers/offline_optimizer.py

```diff
@@ -0,0 +1,139 @@
+                            i += 1
+                            continue
+
+                        i += 1
+                        j += 1
+                        cur_episode = self.cur_batch.build_and_reset()
+                        # Calculate delayed reward
+                        self._calc_delayed_reward(cur_episode)
+                        # Handle everything as if multiagent
+                        if isinstance(cur_episode, SampleBatch):
+                            cur_episode = MultiAgentBatch({
+                                DEFAULT_POLICY_ID: cur_episode
+                            }, cur_episode.count)
+                        for policy_id, s in cur_episode.policy_batches.items():
+                            for row in s.rows():
+                                self.replay_buffers[policy_id].add(
+                                    pack_if_needed(row["obs"]),
+                                    row["actions"],
+                                    row["rewards"],
+                                    pack_if_needed(row["new_obs"]),
+                                    row["dones"],
+                                    weight=None)
+                        self.num_steps_sampled += cur_episode.count
+                else:
+                    self.cur_batch.add_batch(batch)
+
+    def _optimize(self):
+        samples = self._replay()
+
+        with self.grad_timer:
+            info_dict = self.local_evaluator.compute_apply(samples)
+            for policy_id, info in info_dict.items():
+                if "stats" in info:
+                    self.learner_stats[policy_id] = info["stats"]
+                replay_buffer = self.replay_buffers[policy_id]
+                if isinstance(replay_buffer, PrioritizedReplayBuffer):
+                    td_error = info["td_error"]
+                    new_priorities = (
+                        np.abs(td_error) + self.prioritized_replay_eps)
+                    replay_buffer.update_priorities(
+                        samples.policy_batches[policy_id]["batch_indexes"],
+                        new_priorities)
+            self.grad_timer.push_units_processed(samples.count)
+
+        self.num_steps_trained += samples.count
+
+    def _replay(self):
+        samples = {}
+        with self.replay_timer:
+            for policy_id, replay_buffer in self.replay_buffers.items():
+                (obses_t, actions, rewards, obses_tp1,
+                 dones) = replay_buffer.sample(self.train_batch_size)
+                weights = np.ones_like(rewards)
+                batch_indexes = -np.ones_like(rewards)
+                samples[policy_id] = SampleBatch({
+                    "obs": obses_t,
+                    "actions": actions,
+                    "rewards": rewards,
+                    "new_obs": obses_tp1,
+                    "dones": dones,
+                    "weights": weights,
+                    "batch_indexes": batch_indexes
+                })
```

Comment:
Is it possible to use the input API and SyncReplayOptimizer or even SyncSamplesOptimizer instead of this? The functionality looks quite similar.

Reply:
Hmm, I guess the problem is that in the input API right now, postprocess_trajectory() is not called on input batches read from files. So there is no way to compute advantages for the batch if it's not already present. If we allowed this, would that resolve the issue?

Reply:
Happy new year! I just came back. I think `postprocess_trajectory()` works if its `sample_batch` argument is fed with a complete trajectory. In my current implementation, I cache the prefix of current trajectory and compute the advantages once a `done` is encountered. To use `SyncReplayOptimizer` or `SyncSamplesOptimizer` instead, where do you suggest to call `postprocess_trajectory()`? Thanks.

Reply:
You can set the "postprocess_inputs" option here: https://ray.readthedocs.io/en/latest/rllib-offline.html#input-api